[core] screw you objc-runtime

Implement workaround to allow for plugin hotloading AGAIN, as originally planned.
When compiling a plugin, we can append the current time in the executable name to force a fresh reload of the binary inside chunkwms process space.

Actually this change broke the --HEAD brew installation..

```
❯❯❯ ~ brew install chunkwm --HEAD -v                                                                                                                        ⏎
==> Installing chunkwm from crisidev/chunkwm
/usr/bin/sandbox-exec -f /tmp/homebrew20170825-94217-f5zpeo.sb nice /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby -W0 -I /usr/local/Homebrew/Library/Homebrew -- /usr/local/Homebrew/Library/Homebrew/build.rb /usr/local/Homebrew/Library/Taps/crisidev/homebrew-chunkwm/chunkwm.rb --verbose --HEAD
==> Cloning https://github.com/koekeishiya/chunkwm.git
Updating /Users/matbigoi/Library/Caches/Homebrew/chunkwm--git
git config remote.origin.url https://github.com/koekeishiya/chunkwm.git
git config remote.origin.fetch +refs/heads/master:refs/remotes/origin/master
git fetch origin
==> Checking out branch master
git checkout -f master --
Already on 'master'
Your branch is up-to-date with 'origin/master'.
git reset --hard origin/master
HEAD is now at f000a50 core: #204 screw you Apple, we will hotload
==> make install
rm -rf ./bin
mkdir -p ./bin
clang++ src/core/chunkwm.mm -O2 -std=c++11 -Wall -Wno-deprecated -o bin/chunkwm -rdynamic -ldl -lpthread -framework Carbon -framework Cocoa
==> make --directory src/chunkc
rm -rf ./bin
mkdir ./bin
clang chunkc.c -O2 -o bin/chunkc
==> make install --directory src/plugins/tiling
rm -rf ./bin/tiling*
mkdir -p ./bin
clang++ plugin.mm -O2 -std=c++11 -Wall -Wno-deprecated -Wno-writable-strings -o bin/tiling_1503682608.so -shared -fPIC -framework Carbon -framework Cocoa -framework ApplicationServices
rm -f ./../../../plugins/tiling.so
ln -s /private/tmp/chunkwm-20170825-94219-85q8zi/src/plugins/tiling/bin/tiling_1503682608.so ./../../../plugins/tiling.so
ln: ./../../../plugins/tiling.so: No such file or directory
make: *** [bin/tiling] Error 1

==> Formula
Tap: crisidev/chunkwm
Path: /usr/local/Homebrew/Library/Taps/crisidev/homebrew-chunkwm/chunkwm.rb
==> Configuration
HOMEBREW_VERSION: 1.3.1-97-g6d393de
ORIGIN: https://github.com/Homebrew/brew
HEAD: 6d393decfe8099cb049c0ee62c6dc5877e54c0d7
Last commit: 31 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 2c6aca4ceb91cb358aeaa61053143655c66a4eb8
Core tap last commit: 56 minutes ago
HOMEBREW_PREFIX: /usr/local
HOMEBREW_REPOSITORY: /usr/local/Homebrew
HOMEBREW_CELLAR: /usr/local/Cellar
HOMEBREW_BOTTLE_DOMAIN: https://homebrew.bintray.com
CPU: octa-core 64-bit haswell
Homebrew Ruby: 2.0.0-p648
Clang: 8.1 build 802
Git: 2.14.1 => /usr/local/bin/git
Perl: /usr/bin/perl
Python: /usr/local/opt/python/libexec/bin/python => /usr/local/Cellar/python/2.7.13_1/Frameworks/Python.framework/Versions/2.7/bin/python2.7
Ruby: /usr/local/var/rbenv/shims/ruby => /usr/local/var/rbenv/versions/2.3.3/bin/ruby
Java: 1.8.0_141
macOS: 10.12.6-x86_64
Xcode: 8.3.3
CLT: 8.3.2.0.1.1492020469
X11: 2.7.11 => /opt/X11
==> ENV
HOMEBREW_CC: clang
HOMEBREW_CXX: clang++
SDKROOT: /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk
MAKEFLAGS: -j8
CMAKE_PREFIX_PATH: /usr/local
CMAKE_INCLUDE_PATH: /usr/include/libxml2:/System/Library/Frameworks/OpenGL.framework/Versions/Current/Headers
CMAKE_LIBRARY_PATH: /System/Library/Frameworks/OpenGL.framework/Versions/Current/Libraries
MACOSX_DEPLOYMENT_TARGET: 10.12
PKG_CONFIG_LIBDIR: /usr/lib/pkgconfig:/usr/local/Homebrew/Library/Homebrew/os/mac/pkgconfig/10.12
ACLOCAL_PATH: /usr/local/share/aclocal
PATH: /usr/local/Homebrew/Library/Homebrew/shims/super:/usr/local/opt/git/bin:/usr/bin:/bin:/usr/sbin:/sbin

Error: crisidev/chunkwm/chunkwm HEAD-f000a50 did not build
Logs:
     /Users/matbigoi/Library/Logs/Homebrew/chunkwm/00.options.out
     /Users/matbigoi/Library/Logs/Homebrew/chunkwm/01.make
     /Users/matbigoi/Library/Logs/Homebrew/chunkwm/01.make.cc
     /Users/matbigoi/Library/Logs/Homebrew/chunkwm/02.make
     /Users/matbigoi/Library/Logs/Homebrew/chunkwm/02.make.cc
     /Users/matbigoi/Library/Logs/Homebrew/chunkwm/03.make
     /Users/matbigoi/Library/Logs/Homebrew/chunkwm/03.make.cc
If reporting this issue please do so at (not Homebrew/brew or Homebrew/core):
https://github.com/crisidev/homebrew-chunkwm/issues
```

--- a/src/plugins/border/makefile
+++ b/src/plugins/border/makefile
@@ -1,8 +1,10 @@
 BUILD_FLAGS		= -O0 -g -std=c++11 -Wall
-BIN_PATH		= ./../../../plugins
-BUILD_PATH      = ./bin
+BUILD_PATH		= ./../../../plugins
+BINS			= $(BUILD_PATH)/border.so
+DEV_BIN_PATH	= ./../../../plugins
+DEV_BUILD_PATH	= ./bin
+DEV_BINS		= $(DEV_BUILD_PATH)/border
 SRC				= ./plugin.mm
-BINS			= $(BUILD_PATH)/border
 LINK			= -shared -fPIC -framework Carbon -framework Cocoa -framework ApplicationServices
 DIR := ${CURDIR}
 NOW := $(shell date "+%s")
@@ -11,16 +13,24 @@ all: $(BINS)
 
 install: BUILD_FLAGS=-O2 -std=c++11 -Wall
 install: clean $(BINS)
+dev: clean $(DEV_BINS)
 
-.PHONY: all clean install
+.PHONY: all clean install dev
+
+$(DEV_BUILD_PATH):
+	mkdir -p $(DEV_BUILD_PATH)
 
 $(BUILD_PATH):
 	mkdir -p $(BUILD_PATH)
 
 clean:
-	rm -rf $(BUILD_PATH)/border*
-	rm -f $(BIN_PATH)/border.so
+	rm -f $(BUILD_PATH)/border.so
+	rm -rf $(DEV_BUILD_PATH)/border*
+	rm -f $(DEV_BIN_PATH)/border.so
 
-$(BUILD_PATH)/border: $(SRC) | $(BUILD_PATH)
+$(DEV_BUILD_PATH)/border: $(SRC) | $(DEV_BUILD_PATH)
 	clang++ $^ $(BUILD_FLAGS) -o $@_$(NOW).so $(LINK)
-	ln -s $(DIR)/$@_$(NOW).so $(BIN_PATH)/border.so
+	ln -sf $(DIR)/$@_$(NOW).so $(DEV_BIN_PATH)/border.so
+
+$(BUILD_PATH)/border.so: $(SRC) | $(BUILD_PATH)
+	clang++ $^ $(BUILD_FLAGS) -o $@ $(LINK)

--- a/src/plugins/tiling/makefile
+++ b/src/plugins/tiling/makefile
@@ -1,8 +1,10 @@
 BUILD_FLAGS		= -O0 -g -DCHUNKWM_DEBUG -std=c++11 -Wall -Wno-deprecated -Wno-writable-strings
-BIN_PATH		= ./../../../plugins
-BUILD_PATH		= ./bin
+BUILD_PATH		= ./../../../plugins
+BINS			= $(BUILD_PATH)/tiling.so
+DEV_BIN_PATH	= ./../../../plugins
+DEV_BUILD_PATH	= ./bin
+DEV_BINS		= $(DEV_BUILD_PATH)/tiling
 SRC				= ./plugin.mm
-BINS			= $(BUILD_PATH)/tiling
 LINK			= -shared -fPIC -framework Carbon -framework Cocoa -framework ApplicationServices
 DIR := ${CURDIR}
 NOW := $(shell date "+%s")
@@ -11,16 +13,24 @@ all: $(BINS)
 
 install: BUILD_FLAGS=-O2 -std=c++11 -Wall -Wno-deprecated -Wno-writable-strings
 install: clean $(BINS)
+dev: clean $(DEV_BINS)
 
-.PHONY: all clean install
+.PHONY: all clean install dev
+
+$(DEV_BUILD_PATH):
+	mkdir -p $(DEV_BUILD_PATH)
 
 $(BUILD_PATH):
 	mkdir -p $(BUILD_PATH)
 
 clean:
-	rm -rf $(BUILD_PATH)/tiling*
-	rm -f $(BIN_PATH)/tiling.so
+	rm -f $(BUILD_PATH)/tiling.so
+	rm -rf $(DEV_BUILD_PATH)/tiling*
+	rm -f $(DEV_BIN_PATH)/tiling.so
 
-$(BUILD_PATH)/tiling: $(SRC) | $(BUILD_PATH)
+$(DEV_BUILD_PATH)/tiling: $(SRC) | $(DEV_BUILD_PATH)
 	clang++ $^ $(BUILD_FLAGS) -o $@_$(NOW).so $(LINK)
-	ln -s $(DIR)/$@_$(NOW).so $(BIN_PATH)/tiling.so
+	ln -sf $(DIR)/$@_$(NOW).so $(DEV_BIN_PATH)/tiling.so
+
+$(BUILD_PATH)/tiling.so: $(SRC) | $(BUILD_PATH)
+	clang++ $^ $(BUILD_FLAGS) -o $@ $(LINK)

